[Calyx] Add enhancements to common tail elimination for `IfOp`.

cgyurgyik opened the pull request:
This adds a pattern for when `then` and `else` regions are both `ParOp`s. It also takes into consideration the parent region of the `IfOp`, since we don't want to unintentionally change the behavior when pulling out `EnableOp`s. 

I also do not touch anything that's not immediately within the given `ParOp`'s body; doing so may also change the behavior of the program.

The case where the `then` and `else` regions are different, e.g. `SeqOp` and  `ParOp`, should never pull out an `EnableOp`; it will always produce worse code, e.g.
```mlir
calyx.seq {
  calyx.if {
    calyx.par { ... calyx.enable @A }
  } else {
    calyx.seq { ... calyx.enable @A } 
  }
}

// Choice A: Place it in a SeqOp.
calyx.seq {
  calyx.if {
    calyx.par { ... }
  } else {
    calyx.seq { ... } 
  }
  calyx.enable @A // Now, it is no longer parallelized for the `then` region.
}

// Choice B: Place it in a ParOp.
calyx.seq {
  calyx.par {
    calyx.if {
      calyx.par { ... }
    } else {
      calyx.seq { ... } 
    }
    calyx.enable @A  // This is just plain wrong.
  } 
}
```

Closes #1861.

mortbopet commented on lib/Dialect/Calyx/CalyxOps.cpp:
```diff
@@ -1416,53 +1416,111 @@ static EnableOp getLastEnableOp(SeqOp parent) {
-  EnableOp lastElseEnableOp = getLastEnableOp(elseSeqOp);
+  auto &thenControl = ifOp.getThenBody()->front();
+  auto &elseControl = ifOp.getElseBody()->front();
+  if (auto r1 = dyn_cast<ParOp>(thenControl);
```
In this case we essentially have 2 separate canonicalization patterns with disjoint pattern match conditions. Again, it might be personal preference, but i would prefer having two distinct rewrite patterns to capture each of the match (and rewrite) conditions.

mikeurbach replied:
I agree. My rationale for using the simple `hasCanonicalizeMethod` was it is easier when there's just one match and rewrite. But now that we are adding a separate way of matching and rewriting, I think using `hasCanonicalizer` with two separate patterns is preferable. If that makes sense, perhaps a first step would be a NFC to convert from `hasCanonicalizeMethod`->`hasCanonicalizer`, and then this PR would be rebased onto that and add a second pattern.

mortbopet commented on lib/Dialect/Calyx/CalyxOps.cpp:
```diff
@@ -1416,53 +1416,111 @@ static EnableOp getLastEnableOp(SeqOp parent) {
+    llvm::StringMap<EnableOp> A = getAllEnableOpsInImmediateBody(r1),
+                              B = getAllEnableOpsInImmediateBody(r2);
+
+    // Compute the intersection between `A` and `B`.
```
I assume that the intersection has been manually calculated due to the map-like structure of A and B, preventing the use of `std::set_intersection`. Is there not a way to use the keys of each map (vector-like) to then do a set intersection with?

cgyurgyik replied:
Hmm I should try `set_intersection` or LLVM's version again. Before I had something more complicated written.

cgyurgyik replied:
So on second thought, this isn't possible since we essentially want set intersection with side effects. 

We could use the keys of each map to perform a set_intersection, then iterate through `A` and `B` and delete keys with the given `groupName`, but I'm not convinced that's any cleaner than this. WDYT?

mortbopet replied:
If it'll complicating the code just for the sake of shoehorning an stl function in there, then that probably doesn't make that much sense, so up to you.

cgyurgyik replied:
Given this is only used once, I don't think its necessary. Perhaps in the future we can right some wrapper for set intersection with side effects. Is there anything else blocking your approval?

mortbopet commented on lib/Dialect/Calyx/CalyxOps.cpp:
```diff
@@ -1416,53 +1416,111 @@ static EnableOp getLastEnableOp(SeqOp parent) {
+    // Compute the intersection between `A` and `B`.
+    SmallVector<StringRef> groupNames;
+    for (auto a = A.begin(); a != A.end(); ++a) {
+      StringRef groupName = a->getValue().groupName();
```
`A` is a mapping of `std::pair(op.groupName(), op)`. Does this not mean that `a->getKey() == a->getValue().groupName())`?

cgyurgyik replied:
Yeah you're probably right I'll fix this.

mortbopet commented on lib/Dialect/Calyx/CalyxOps.cpp:
```diff
@@ -1416,53 +1416,111 @@ static EnableOp getLastEnableOp(SeqOp parent) {
+    ///       calyx.enable @A     ->         par { ... }
+    ///       calyx.enable @B              }
+    ///     }                              calyx.enable @A
+    ///   else {                           calyx.enable @B
```
What is the context that `calyx.enable @A, calyx.enable @B` gets extracted to? Are we expecting the `if` to be nested inside a `seq`, and `calyx.enable @A, calyx.enable @B` to be placed within a `par` following the `if``?

cgyurgyik replied:
If the `then`, `else` regions are nested in `seq`s, then we should nest the `if` and extracted `enable`s in a `seq` as well. Similar for `par`. This ensures we are still maintaining similar code behavior. In the worst case, we have `seq { seq { ... } }`, but this gets canonicalized away. Did that answer your question?

mortbopet replied:
I understand; my main concern was just that it wasn't included in the code example, i.e. groups going from clearly being within a `par` group to be extracted into the parent group, which is not shown.

cgyurgyik replied:
What do you mean by "parent group"?

The example doesn't show the `IfOp`'s parent mostly because there are two different cases, and the behavior is somewhat expected, e.g. if an `EnableOp` is driven sequentially, it will still be driven sequentially when pulled out.

If you think this isn't clear, I can add the parent to the example.

mortbopet commented on lib/Dialect/Calyx/CalyxOps.cpp:
```diff
@@ -1416,53 +1416,111 @@ static EnableOp getLastEnableOp(SeqOp parent) {
-  rewriter.eraseOp(lastThenEnableOp);
-  rewriter.eraseOp(lastElseEnableOp);
-  return success();
+  return failure();
```
Again, as above, i think it would be more clear if these two canonicalizations are split out into two rewrite patterns, and this pattern match failure statement is placed as early as possible.

mortbopet commented:
Whether it is possible to indicate in the comment what happens when the parent group of the `IfOp` is a `SeqOp` vs a `ParOp`, i.e.:
```
par {                         par {
  if %a with @G {               if %a with @G {
    par {                         par { ... }
      ...                       } else {
      calyx.enable @A     ->      par { ... }
      calyx.enable @B           }
    }                           calyx.enable @A
  else {                        calyx.enable @B
    par {                     }
      ...
      calyx.enable @A
      calyx.enable @B
    }
  }
}
```
or with a new `par` op:
```
seq {                         seq {
  if %a with @G {               par {
    par {                         if %a with @G {
      ...                           par { ... }
      calyx.enable @A     ->      } else {
      calyx.enable @B               par { ... }
    }                             }
  else {                          calyx.enable @A
    par {                         calyx.enable @B
      ...                       }
      calyx.enable @A           ...
      calyx.enable @B         }
    }
  }
  ...
}
```
As far as i can tell, the latter is always the case (a new `par` op is created); canonicalization will then remove any nested `par` ops if the parent was not a `seq`.

I think it would be fine to just modify the comment to the following, to indicate that _no matter what_ things will get nested inside a new `par` op.
```
  if %a with @G {              par {
    par {                        if %a with @G {
      ...                          par { ... }
      calyx.enable @A            } else {
      calyx.enable @B     ->       par { ... }
    }                            }
  else {                         calyx.enable @A
    par {                        calyx.enable @B
      ...                      }
      calyx.enable @A
      calyx.enable @B
    }  
  }  
```

rachitnigam commented:
Just a drive-by comment to say this looks super exciting! This pass doesn't exist in the Calyx infrastructure so it's a completely new Calyx optimization!

cgyurgyik commented:
@mortbopet Your code examples are correct. Cool, I agree with your example and I'll update it.

cgyurgyik commented:
> Just a drive-by comment to say this looks super exciting! This pass doesn't exist in the Calyx infrastructure so it's a completely new Calyx optimization!

Credits to @mortbopet for taking the initiative on this one.

@mortbopet, @mikeurbach  
b339274 addresses nits, a8d9aba changes it to `hasCanonicalizer` for the `IfOp`.

cgyurgyik commented:
In #1883, I deleted the canonicalizations for empty regions. I think I was mistaken here, since I wasn't getting the results I expected. Part of this is because these canonicalizations were not taking into consideration arguments. For example, erasing an `IfOp` may lead to the invalidation of certain conditions, such as an unused group. I'll address this in a follow-up PR.

cgyurgyik commented:
Going to push this for now; I will address any comments post-commit.

mikeurbach commented:
Thanks for pulling out two separate patterns, that was my only feedback.
